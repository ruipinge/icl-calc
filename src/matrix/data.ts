import { DATA_POINTS, DataPoint } from '../db';

const ATA_DEVIATION: number = 0.1; // millimetres
const CLR_DEVIATION: number = 100 * 0.001; // millimetres

export const NUM_DATA_POINTS = DATA_POINTS.length;

export interface VaultRange {
  min?: number;
  max?: number;
}

type LensSizeId = 'small' | 'medium' | 'large';
interface LensSize {
  id: LensSizeId;
  label: string;
  value: number;
}
export const LENS_SIZES: LensSize[] = [
  {
    label: '12.6 mm',
    id: 'small',
    value: 12.6
  },
  {
    label: '13.2 mm',
    id: 'medium',
    value: 13.2
  },
  {
    label: '13.7 mm',
    id: 'large',
    value: 13.7
  }
];

type MyopiaId = 'low' | 'moderate' | 'high';
interface MyopiaLevel {
  id: MyopiaId;
  label: string;
  title: string;
  min: number;
  max: number;
}
export const MYOPIA_LEVELS: MyopiaLevel[] = [
  {
    id: 'low',
    label: 'Low',
    title: 'Spherical Equivalent (SE) less than -6 dpt',
    min: -6,
    max: Infinity
  },
  {
    id: 'moderate',
    label: 'Moderate',
    title: 'Spherical Equivalent (SE) between -6 dpt and -12 dpt',
    min: -12,
    max: -6
  },
  {
    id: 'high',
    label: 'High',
    title: 'Spherical Equivalent (SE) greater than -12 dpt',
    min: -Infinity,
    max: -12
  }
];

export interface MatrixFilter {
  readonly ata: number;
  readonly clr: number;
  readonly dataPoints?: DataPoint[];
}

interface CacheEntry {
  lens: LensSize;
  myopia: MyopiaLevel;
  dataPoints?: DataPoint[];
}

export const getDataPoints = ({
  lens,
  myopia,
  dataPoints = DATA_POINTS
}: CacheEntry): DataPoint[] =>
  dataPoints.filter(
    (point) =>
      lens.value === point.iclSize &&
      myopia.min <= point.iclSe &&
      point.iclSe < myopia.max
  );

const filter = ({
  ata,
  clr,
  point
}: {
  ata: number;
  clr: number;
  point: DataPoint;
}) =>
  ata >= point.ata - ATA_DEVIATION &&
  ata <= point.ata + ATA_DEVIATION &&
  clr >= point.clr - CLR_DEVIATION && // clr in db is in millimetres
  clr <= point.clr + CLR_DEVIATION; // clr in db is in millimetres

export const filterFlatRows = ({
  ata,
  clr,
  dataPoints = DATA_POINTS
}: MatrixFilter): DataPoint[][] =>
  LENS_SIZES.reduce(
    (memoLens, lens) => [
      ...memoLens,
      ...MYOPIA_LEVELS.reduce(
        (memoMyopia, level) => [
          ...memoMyopia,
          ...[
            getDataPoints({
              lens: lens,
              myopia: level,
              dataPoints: dataPoints
            }).filter((point: DataPoint) => filter({ ata, clr, point }))
          ]
        ],
        [] as DataPoint[][]
      )
    ],
    [] as DataPoint[][]
  );
