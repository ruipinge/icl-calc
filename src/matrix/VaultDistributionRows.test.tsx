import {
  VaultDistributionRows,
  countByVaultRange,
  formatVaultSizeTexts,
  getVaultDistribution
} from './VaultDistributionRows';

import TestRenderer from 'react-test-renderer';
import { FILTER as filter } from './data.test';

it('renders without crashing', () => {
  const tree = TestRenderer.create(
    <VaultDistributionRows ata={0} clr={0} />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('calculates vault size distribution with max', () => {
  const row = getVaultDistribution({ filter, range: { max: 1.5 } });
  expect(row).toEqual([1, 0.333, 0, 0, 0, 0, 0, 0, 0]);
});

it('calculates vault size distribution with min', () => {
  const row = getVaultDistribution({ filter, range: { min: 3.4 } });
  expect(row).toEqual([0, 0, 0, 0, 0, 0, 0, 0.333, 1]);
});

it('calculates vault size distribution with min and max', () => {
  const row = getVaultDistribution({
    filter,
    range: { min: 2.2, max: 2.7 }
  });
  expect(row).toEqual([0, 0, 0, 0.333, 1, 0.333, 0, 0, 0]);
});

it('formats vault size texts without range', () => {
  const texts = formatVaultSizeTexts({});
  expect(texts).toEqual({ label: '', title: '' });
});

it('formats vault size texts with min', () => {
  const texts = formatVaultSizeTexts({ min: 0 });
  expect(texts).toEqual({
    title: 'Percentage of Eyes with Vault size greater or equal than 0 mm',
    label: '% 0 < Vault'
  });
});

it('formats vault size texts with max', () => {
  const texts = formatVaultSizeTexts({ max: 0 });
  expect(texts).toEqual({
    title: 'Percentage of Eyes with Vault size less than 0 mm',
    label: '% Vault < 0'
  });
});

it('formats vault size texts with min and max', () => {
  const texts = formatVaultSizeTexts({ max: 10, min: 0 });
  expect(texts).toEqual({
    title: 'Percentage of Eyes with Vault size between 0 mm and 10 mm',
    label: '% 0 < Vault < 10'
  });
});

it('counts rows by vault range', () => {
  const a = countByVaultRange({ rows: [[1, 2]], range: {} });
  expect(a).toBe(0);
});
