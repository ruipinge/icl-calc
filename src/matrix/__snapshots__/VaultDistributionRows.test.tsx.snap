// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
Array [
  <tr>
    <th
      scope="row"
      title="Percentage of Eyes with Vault size less than 250 micrometres"
    >
      % Vault &lt; 250 (μm)
    </th>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
  </tr>,
  <tr>
    <th
      scope="row"
      title="Percentage of Eyes with Vault size between 250 and 500 micrometres"
    >
      % 250 &lt; Vault &lt; 500 (μm)
    </th>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
  </tr>,
  <tr>
    <th
      scope="row"
      title="Percentage of Eyes with Vault size between 500 and 750 micrometres"
    >
      % 500 &lt; Vault &lt; 750 (μm)
    </th>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
  </tr>,
  <tr>
    <th
      scope="row"
      title="Percentage of Eyes with Vault size between 750 and 1000 micrometres"
    >
      % 750 &lt; Vault &lt; 1000 (μm)
    </th>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
  </tr>,
  <tr>
    <th
      scope="row"
      title="Percentage of Eyes with Vault size greater or equal than 1000 micrometres"
    >
      % 1000 &lt; Vault (μm)
    </th>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
    <td
      className="text-right"
    >
      0
    </td>
  </tr>,
]
`;
